lint/clang_tidy.bzl

```diff
@@ -0,0 +1,353 @@
+"""API for calling declaring a clang-tidy lint aspect.
+
+Typical usage:
+
+First, install clang-tidy with llvm_toolchain or as a native binary (llvm_toolchain
+does not support Windows as of 06/2024, but providing a native clang-tidy.exe works)
+
+Next, declare a binary target for it, typically in `tools/lint/BUILD.bazel`:
+
+e.g. using llvm_toolchain:
+```starlark
+native_binary(
+    name = "clang_tidy",
+    src = "@llvm_toolchain_llvm//:bin/clang-tidy"
+    out = "clang_tidy",
+)
+```
+
+e.g as native binary:
+```starlark
+native_binary(
+    name = "clang_tidy",
+    src = "clang-tidy.exe"
+    out = "clang_tidy",
+)
+```
+
+Finally, create the linter aspect, typically in `tools/lint/linters.bzl`:
+
+```starlark
+load("@aspect_rules_lint//lint:clang_tidy.bzl", "clang_tidy_aspect")
+
+clang_tidy = clang_tidy_aspect(
+    binary = "@@//path/to:clang-tidy",
+    configs = "@@//path/to:.clang-tidy",
+)
+```
+"""
+
+load("@aspect_bazel_lib//lib:copy_to_bin.bzl", "COPY_FILE_TO_BIN_TOOLCHAINS", "copy_files_to_bin_actions")
+load("@aspect_rules_js//js:libs.bzl", "js_lib_helpers")
+load("@bazel_tools//tools/build_defs/cc:action_names.bzl", "ACTION_NAMES")
+load("@bazel_tools//tools/cpp:toolchain_utils.bzl", "find_cpp_toolchain")
+load("//lint/private:lint_aspect.bzl", "LintOptionsInfo", "dummy_successful_lint_action", "filter_srcs", "report_files", "patch_and_report_files")
+
+_MNEMONIC = "AspectRulesLintClangTidy"
+
+###############################################
+# START COPY CODE FROM private/lint_aspect.bzl
+#
+# this code needed to be changed to support generation of a report file
```

Comment:
Interesting, why is clang-tidy the only tool that needs to produce different number of reports? I think this will break the API this ruleset exposes to surrounding tools that want to present the results - they assume there is a report for each action. Maybe we can merge them?